src/box/iproto.cc

```diff
@@ -1720,6 +1808,79 @@ tx_inject_delay(void)
 	});
 }
 
+static void
+tx_process_begin(struct cmsg *m)
+{
+	struct iproto_msg *msg = tx_accept_msg(m);
+	struct obuf *out;
+	struct iproto_stream *stream = msg->stream;
+
+	if (tx_check_schema(msg->header.schema_version))
+		goto error;
+
+	if (box_txn_begin() != 0)
+		goto error;
+
+	stream->txn = in_txn();
+	txn_detach();
+
+	out = msg->connection->tx.p_obuf;
+	iproto_reply_ok(out, msg->header.sync, ::schema_version);
+	iproto_wpos_create(&msg->wpos, out);
+	return;
+error:
+	tx_reply_error(msg);
+	tx_end_msg();
+}
+
+static void
+tx_process_commit(struct cmsg *m)
+{
+	struct iproto_msg *msg = tx_accept_msg(m);
+	struct obuf *out;
+	struct iproto_stream *stream = msg->stream;
+
+	if (tx_check_schema(msg->header.schema_version))
+		goto error;
+
+	if (box_txn_commit() != 0) {
+		stream->txn = in_txn();
```

Comment:
Why do you need it? Isn't `stream->txn` now equal to `in_txn()` anyway?

Reply:
After box_txn_commit fails in_txn() can return 'NULL' or pointer to transaction in case if (txn->in_sub_stmt)